{% extends "template.html" %} {% block content %}
<div class="about">
  <h1>About</h1>
  <div class="content-container">
    <section>
      <h3>Modules</h3>
      <p>
        Below is the entire process of sequence generation, prediction, and
        evaluation, along with the models used.
      </p>
      <img
        src="{{ url_for('static', filename='img/modules.png') }}"
        alt="Modules"
      />
    </section>
  
    <section>
      <h3>Modles</h3>
      <p>References of all models used in entire process.</p>
      <table>
        <thead>
          <tr>
            <th>Model</th>
            <th>Details</th>
            <th>Reference</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td><strong><a href="https://www.ebi.ac.uk/jdispatcher/msa/clustalo">Clustal Omega</a></strong></td>
            <td>
              A version, completely rewritten and revised in 2011, of the widely
              used Clustal series of programs for multiple sequence alignment.
            </td>
            <td>
              Sievers, Fabian and Desmond G. Higgins. "The Clustal Omega Multiple
              Alignment Package." Methods in molecular biology 2231 (2021): 3-16 .
            </td>
          </tr>
          <tr>
            <td><strong><a href="https://github.com/microsoft/evodiff">EvoDiff</a></strong></td>
            <td>
              Able to generate diverse protein sequences in both conditional and
              unconditional states, predict their structures, and use OmegaFold to
              evaluate the quality.
            </td>
            <td>
              Alamdari, Sarah, Nitya Thakkar, Rianne van den Berg, Alex X. Lu,
              Nicolo Fusi, Ava P. Amini and Kevin Kaichuang Yang. "Protein
              generation with evolutionary diffusion: sequence is all you need."
              bioRxiv (2023): n. pag.
            </td>
          </tr>
          <tr>
            <td><strong><a href="https://github.com/HeliXonProtein/OmegaFold">OmegaFold</a></strong></td>
            <td>Predict structure of protein sequence</td>
            <td>
              Wu, Rui Min, Fan Ding, Rui Wang, Rui Shen, Xiwen Zhang, Shitong Luo,
              Chenpeng Su, Zuofan Wu, Qi Xie, Bonnie Berger, Jianzhu Ma and Jian
              Peng. "High-resolution de novo structure prediction from primary
              sequence." bioRxiv (2022): n. pag.
            </td>
          </tr>
          <tr>
            <td><strong><a href="https://interproscan-docs.readthedocs.io/en/latest/">InterProScan</a></strong></td>
            <td>
              <ul>
                <li>Scan motif for sequence.</li>
                <li>Released on 25 July 2024: InterProScan 5.69-101.0</li>
              </ul>
            </td>
            <td>
              Jones, Philip, David Binns, Hsin-Yu Chang, Matthew Fraser, Weizhong
              Li, Craig McAnulla, Hamish McWilliam, John Maslen, Alex L. Mitchell,
              Gift Nuka, Sebastien Pesseat, Antony F. Quinn, Amaia
              Sangrador-Vegas, Maxim Scheremetjew, Siew-Yit Yong, Rodrigo Lopez
              and Sarah Hunter. "InterProScan 5: genome-scale protein function
              classification." Bioinformatics 30 (2014): 1236 - 1240.
            </td>
          </tr>
          <tr>
            <td><strong><a href="https://github.com/ievapudz/TemStaPro">TemStaPro</a></strong></td>
            <td>
              Predict protein thermostability using embeddings generated by
              protein language models (pLMs) from an input protein sequence.
            </td>
            <td>
              Pudžiuvelytė, Ieva, Kliment Olechnovič, Eglė Godliauskaitė,
              Kristupas Šermokas, Tomas Urbaitis, Giedrius Gasiunas and Darius
              Kazlauskas. "TemStaPro: protein thermostability prediction using
              sequence representations from protein language models."
              Bioinformatics 40 (2024): n. pag.
            </td>
          </tr>
          <tr>
            <td><strong><a href="https://github.com/beckham-lab/EpHod">EpHod</a></strong></td>
            <td>
              Learns structural and biophysical features from sequence data alone
              to predict enzyme pH<sub>opt</sub> from protein sequence.
            </td>
            <td>
              Gado, Japheth E., Matthew Knotts, Ada Y. Shaw, Debora S. Marks,
              Nicholas Paul Gauthier, Christensen Carsten Sander and Gregg T.
              Beckham. "Deep learning prediction of enzyme optimum pH." bioRxiv
              (2023): n. pag.
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>
</div>


{% endblock %}
